File: leetcode-java/src/main/java/chapter02/PalindromeList.java
package chapter02;

import utils.ListNode;

import java.util.LinkedList;

/**
 * <p>回文链表</p>
 */
public class PalindromeList {

    /**
     * <p>思路: 栈实现</p>
     */
    public static boolean isPalindromeStack(ListNode head) {
        ListNode current = head;
        LinkedList<Integer> stack = new LinkedList<>();
        while (current != null) {
            stack.push(current.value);
            current = current.next;
        }
        current = head;
        while (current != null) {
            if (current.value != stack.pop()) {
                return false;
            }
            current = current.next;
        }
        return true;
    }

    /**
     * <p>思路: 栈实现, 优化</p>
     */
    public static boolean isPalindromeStackOptimize(ListNode head) {
        boolean flag = false;
        ListNode current = head;
        ListNode slow = head, fast = head;
        LinkedList<Integer> stack = new LinkedList<>();
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        while (current != slow) {
            stack.push(current.value);
            current = current.next;
        }
        if (flag) {
            stack.push(current.value);
        }
        while (current != null) {
            if (!stack.isEmpty() && current.value != stack.pop()) {
                return false;
            }
            current = current.next;
        }
        return true;
    }

    /**
     * <p>思路：</p>
     * <p>1. 快慢指针找到中点, 然后从中点开始逆序直到结尾</p>
     * <p>2. 快指针和慢指针两个分别从头和尾开始遍历, 直到中点</p>
     */
    public static boolean isPalindrome(ListNode head){
        ListNode fast = head;
        ListNode slow = head;
        // 快慢指针找中点
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode first = head;
        ListNode second = ReverseList.reverseListLoop(slow);
        while (second != null) {
            if (first.value != second.value) {
                return false;
            }
            first = first.next;
            second = second.next;
        }
        return true;
    }

}

File: leetcode-go/chapter02/palindrome_list.go
package main

import "leetcode-go/node"

type ListPalindromer struct {

}

// 回文链表
func (palindromer *ListPalindromer) isPalindrome(head *node.ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	fast = head
	reverser := &ListReverser{}
	slow = reverser.ReverseListLoop(slow)
	for slow != nil {
		if slow.Val != fast.Val {
			return false
		}
		slow = slow.Next
		fast = fast.Next
	}
	return true
}

// 回文链表: 数组 + 双指针
func isPalindromeV1(head *node.ListNode) bool {
	// 1. 遍历链表
	stack := make([]*node.ListNode, 0)
	for head != nil {
		stack = append(stack, head)
		head = head.Next
	}
	// 2. 双指针遍历
	first, second := 0, len(stack)-1
	for first <= second {
		if stack[first].Val != stack[second].Val {
			return false
		}
		first++
		second--
	}
	return true
}

// 回文链表: 快慢指针 + 反转链表
func isPalindromeV2(head *node.ListNode) bool {
	slow, fast := head, head
	// 1. 快慢指针找中点
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	// 2. 反转链表
	var current *node.ListNode = slow
	var previous *node.ListNode = nil
	for current != nil {
		next := current.Next
		current.Next = previous
		previous = current
		current = next
	}
	// 3. 遍历链表对比
	slow, fast = head, previous
	for fast != nil {
		if slow.Val != fast.Val {
			return false
		}
		slow, fast = slow.Next, fast.Next
	}
	return true
}
